File: shop/src/main/java/basic/shop/order/command/domain/entiy/Order.java
package basic.shop.order.command.domain.entiy;

import basic.shop.common.model.AbstractBaseEntity;
import basic.shop.common.model.Money;
import basic.shop.order.command.domain.value.OrderNo;
import basic.shop.order.command.domain.value.OrderState;
import basic.shop.order.command.domain.value.ShippingInfo;
import lombok.Getter;
import lombok.ToString;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Entity
@Table(name = "orders")
@Access(AccessType.FIELD)
@Getter
@ToString(of = {"id", "orderLines", "totalAmounts", "shippingInfo", "orderState"})
public class Order extends AbstractBaseEntity {
    @EmbeddedId
    private OrderNo id;
    private List<OrderLine> orderLines = new ArrayList<>();
    private Money totalAmounts;
    private ShippingInfo shippingInfo;
    private OrderState orderState;

    public Order(List<OrderLine> orderLines, ShippingInfo shippingInfo) {
        setOrderLines(orderLines);
        setShippingInfo(shippingInfo);
    }

    public void changeShippingInfo(ShippingInfo shippingInfo) {
        verifyNotYetShipped();
        this.shippingInfo = shippingInfo;
    }

    public void cancel() {
        verifyNotYetShipped();
        this.orderState = OrderState.CANCELED;
    }

    private void setOrderLines(List<OrderLine> orderLines) {
        verifyAtLeastOneOrMoreOrderLines(orderLines);
        this.orderLines = orderLines;
        calculateTotalAmounts();
    }

    private void verifyAtLeastOneOrMoreOrderLines(List<OrderLine> orderLines) {
        if (orderLines == null || orderLines.isEmpty()) {
            throw new IllegalArgumentException("no OrderLine");
        }
    }

    private void calculateTotalAmounts() {
        int sum = orderLines.stream().mapToInt(OrderLine::getAmounts).sum();
        this.totalAmounts = new Money(sum);
    }

    private void setShippingInfo(ShippingInfo shippingInfo) {
        try {
            Objects.requireNonNull(shippingInfo, "no ShippingInfo");
        } catch(NullPointerException e) {
            throw new IllegalArgumentException(e.getMessage());
        }
        this.shippingInfo = shippingInfo;
    }

    private void verifyNotYetShipped() {
        if(orderState != OrderState.PAYMENT_WAITING && orderState != OrderState.PREPARING) {
            throw new IllegalArgumentException("already Shipped");
        }
    }

    @Override
    public boolean equals(Object o) {
        if(o == this)
            return true;
        if(!(o instanceof Order order))
            return false;
        return Objects.equals(id, order.id) && isOrderLines(order.orderLines)
                && Objects.equals(totalAmounts, order.totalAmounts) && Objects.equals(shippingInfo, order.shippingInfo);
    }

    private boolean isOrderLines(List<OrderLine> orderLines) {
        int arrayCount = this.orderLines.size();
        int compareCount = (int) this.orderLines.stream().filter(orderLines::contains).count();
        return arrayCount == compareCount;
    }

    @Override
    public int hashCode() {
        int result = id == null ? 0 : id.hashCode();
        result = result * 31 + orderLines.hashCode();
        result = result * 31 + totalAmounts.hashCode();
        result = result * 31 + shippingInfo.hashCode();
        return result;
    }
}

File: spring-mvc-with-kotlin/order-api/src/main/kotlin/basic/orderapi/order/command/domain/entity/Order.kt
package basic.orderapi.order.command.domain.entity

import basic.orderapi.common.jpa.MoneyConverter
import basic.orderapi.common.model.AbstractTimeEntity
import basic.orderapi.common.model.Address
import basic.orderapi.common.model.Money
import basic.orderapi.order.command.domain.value.OrderLine
import basic.orderapi.order.command.domain.value.OrderNo
import basic.orderapi.order.command.domain.value.OrderState
import basic.orderapi.order.command.domain.value.Orderer
import javax.persistence.*

@Entity
@Table(name = "purchase_order")
@Access(AccessType.FIELD)
class Order (
    @EmbeddedId
    var orderNumber: OrderNo?,

    @ElementCollection(fetch = FetchType.LAZY)
    @CollectionTable(name = "order_line", joinColumns = [JoinColumn(name = "order_number")])
    var orderLines: MutableList<OrderLine>,

    @Convert(converter = MoneyConverter::class)
    var totalAmounts: Money,

    @Enumerated(EnumType.STRING)
    var orderState: OrderState,

    @Embedded
    var shippingInfo: Address,

    @Embedded
    var orderer: Orderer
): AbstractTimeEntity() {
    companion object {
          fun generate(orderNumber: OrderNo, orderLines: List<OrderLine>,
                       shippingInfo: Address, orderState: OrderState, orderer: Orderer): Order {
              verifyAtLeastOneOrMoreOrderLInes(orderLines)
              return Order(orderNumber, orderLines.toMutableList(), calculateTotalAmounts(orderLines),
                  orderState, shippingInfo, orderer)
          }
        private fun verifyAtLeastOneOrMoreOrderLInes(orderLines: List<OrderLine>) {
            if(orderLines.isNullOrEmpty())
                throw IllegalArgumentException("No OrderLine")
        }
        private fun calculateTotalAmounts(orderLines: List<OrderLine>): Money {
            return Money(orderLines.sumOf { it.amounts.value })
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Order

        if (orderNumber != other.orderNumber) return false
        if (shippingInfo != other.shippingInfo) return false
        if (orderLines != other.orderLines) return false
        if (totalAmounts != other.totalAmounts) return false
        if (orderState != other.orderState) return false

        return true
    }

    override fun hashCode(): Int {
        var result = orderNumber?.hashCode() ?: 0
        result = 31 * result + shippingInfo.hashCode()
        result = 31 * result + orderLines.hashCode()
        result = 31 * result + totalAmounts.hashCode()
        result = 31 * result + orderState.hashCode()
        return result
    }
}